contracts/fate.md

```diff
@@ -0,0 +1,648 @@
+* Difficulty
+* Gaslimit
+* Address
+* Balance
+* Caller
+* Value
+
+#### The contract store
+
+The permanent storage of a contract. Stored in the chain state
+tree. This is a key value store that is mapped to the contract state
+by the compiler.
+
+The compiler can choose how to represent the contract state in the
+store. For instance, save the entire state under a single key (this is
+how it works in AEVM at the moment), or it could flatten any state
+records and store each field under a separate key.
+
+#### The state tree
+
+Contains all accounts, oracles and contracts on the chain. Most values
+such as account balances can be read directly through specific load
+instructions. Some values can be changed through transaction
+instructions.
+
+#### The local storage
+
+The local storage is a stack of key value storages. That is, a list of
+maps. When reading a key, the machine first looks in the first map of
+the list. If a key is found and the type is correct the value is
+returned. If the type is incorrect an exception is thrown. If no value
+is found in the first map, the machine will look in the next map in
+the list and so on. If the key does not exist in any of the maps an
+exception is thrown.
+
+The key is a an integer (32 bit).
+
+#### The accumulator stack
+
+The accumulator stack is a stack of typed values. The top of the stack
+can be used as an argument to an operation. Values can be pushed to
+the stack and popped from the stack.
+
+#### The execution stack
+
+The execution stack contain return addresses as a tuple in the form of
+{contract address, function hash, and basic block index}.
+
+#### The event stream
+
+A contract can emit events similar to the EVM.
+
+## Types
+
+The machine supports the following types:
+* Integer (signed arbitrary precision integers) [I]
+* Boolean (true | false) [B]
+* Addresses (A pointer into the state tree) [A]
+* Strings (utf8 encoded byte arrays) [S]
+* Tuples (), ('a), ('a, 'b, ...) [T]
+* Lists: Cons cells (‘a) | Nil (Ʇ)  [L],[C],[Ʇ]
+* Maps (‘a, ‘b),  (Key: ‘a -> Value: ‘b)  ‘a not a map [M]
+* Variant Types, (| Size | Tag | Elements  |)  [V]
+* Bits <Bits> or !<Bits>
+* TypeRep [R]
+
+These types can be used as argument and return types of functions and
+also as the type of the elements in the contract store.
+
+### Integer
+
+The type Integer, is used for any integer, infinitely large or small
+(below zero).
+
+Examples:
+
+```
+ 0
+ 1
+-1
+589465789334
+-51315353513513131656462262
+```
+
+### Boolean
+
+The Boolean type only has two values, true or false.
+
+Examples (all of them):
+```
+ true
+ false
+```
+
+Operations on booleans include:
+*   and, or, not
+*   conditional jumps
+
+### Addresses
+
+Values of the type Address are 32-bytes (256-bits) pointers to
+entities on the chain (Accounts, Oracles, Contracts, etc)
+
+Examples:
+```
+<<ak_2HNb8bivUoJTcaxD6VKDy2wPHvTuQgQRJ3dkF9RSyVFqBkMneC>>
+```
+
+### Strings
+
+Strings are byte arrays od UTF-8 encoded unicode characters. (These
```

Comment:
Re UTF-8: is it enforced in the language or in the VM?

I expect that at the lowest level the string is represented as a sequence of bytes. But not all sequence of bytes are valid UTF-8:
```
$ erl
Erlang/OTP 20 [erts-9.3.3.9] [source] [64-bit] ...
$ erl
Erlang/OTP 20 [erts-9.3.3.9] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:10] [hipe] [kernel-poll:false]
Eshell V9.3.3.9  (abort with ^G)
1> lists:map(fun unicode:characters_to_list/1, [<<2#00100100>>, <<2#11000010, 2#10100010>>]).
["$","¢"]
2> lists:map(fun unicode:characters_to_list/1, [<<2#10000000>>, <<2#10000010, 2#10100010>>]).
[{error,[],<<128>>},{error,[],<<130,162>>}]
```

Does it mean that all points where strings are instantiated in the VM are checked to be valid UTF-8? Or rather an exception is raised by the VM when the assumption of UTF-8 validity is enforced?

Also, I need to understand better if the validity of a string according to UTF-8 changes depending on the version of Unicode.